### DISCUSSION TITLE
Alt Text Character Limit

### DISCUSSION THREAD

**Discussion by cdesrosiers813** [w3c/wcag#4047] [Ideas (WCAG 2.x / WCAG 3)]
I created this issue on behalf of the [Alt Text Subgroup] (https://www.w3.org/WAI/GL/task-forces/silver/wiki/Alt_Text_Subgroup) of the Silver Task Force. 

Explanation:
Up until this point, W3C has not specified a character limit for alt text. The guidelines refer to “short text alternative” and “long text alternative.” This is leading to confusion and mixed messages on the web with organizations attempting to define this limit for users: 

[University of Illinois](https://fae.disability.illinois.edu/rulesets/IMAGE_4_EN/): 100 characters
[Moz](https://moz.com/learn/seo/alt-text#:~:text=Alt%20text%20(alternative%20text)%2C,an%20image%20on%20a%20page.): 125 characters
[HubSpot](https://blog.hubspot.com/marketing/image-alt-text): 125 characters
[GVSU](https://www.gvsu.edu/accessibility/alt-text-7.htm): 140 characters
[Aces Editing](https://aceseditors.org/news/2020/how-to-write-great-alt-text-and-why-it-matters): 140 characters

This topic has been discussed within W3C in the past, and some have raised concerns with defining a character limit. First, restricting the character limit may discourage people from providing a sufficient text alternative for an image. Second, many platforms do not support extended description. Third, the same alt text can have different character lengths when written in different languages. 

Question:

Why doesn’t WCAG 2.x set a character limit for the alt attribute value?

**Comment by BitwiseAndrea** [comment]
Twitters implementation of alt text allows up to 1k characters and is also visible to users who do not use screenreaders behind a button. When seeing this I wondered why longdesc wasn't used instead. If an alt text max length was to be added it seems like it would be a reasonable thing to suggest using longdesc in cases where more than 100 characters were necessary, and improving the pitfalls longdesc.

**Comment by fstrr** [comment]
The `longdesc` attribute isn't part of the HTML specification anymore and it has inconsistent support from browsers and assistive technology, which could be why Twitter didn't use it. [The Accessibility Guidelines Working Group is proposing to remove the related H45 technique](https://github.com/w3c/wcag/issues/2502). Alt text doesn't have to fit into 100 characters—the [Non-Text Content criterion](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html) requires that a text alternative serves an equivalent purpose, and that could lead to a longer description.

**Comment by cstrobbe** [comment]
Authors should consider two things:

1. How screen readers deal with "long" `alt` attributes: screen readers may split the contents into chunks of 125 characters. (This has led some people to believe, mistakenly, that the `alt` attribute gets truncated at 125 characters.)
2. As a screen reader user, you [can't pause the `alt` attribute](https://yatil.net/blog/there-is-no-character-limit-for-alt-text#komment_566edcf224bd7283ecee75626aa82acb); you listen to all of it or start over. For this reason, a long text alternative can be frustrating.

However, the relationship between the number of characters on the one hand and the number of spoken syllables (in speech synthesis) on the other varies from language to language (and from writing system to writing system). A general character limit would be unable to take this variability into account. As a consequence, any character limit for the `alt` attribute would be perceived as arbitrary.

**Comment by GreggVan** [comment]
hmmmm

But without long desc in HTML (and not universally supported)   and aria not having an equivalent (other than pointing to some other url) what is our recommendation for  if there is a picture that has more information than can be expressed in short ALT?     (Also we have Powerpoints and other formats that only have ALT text.

Is our advice — make it as long as necessary to convey all important information — but not a word longer?


Gregg Vanderheiden
***@***.***

**Comment by cstrobbe** [comment]
I'm not questioning that there needs to be guidance to help authors decide when they should just use an `alt` attribute (in HTML) or consider complementing that with a long description that is presented through some other means. (The [WAI tutorial on Complex Images](https://www.w3.org/WAI/tutorials/images/complex/) provides a number of methods without being exhaustive.)

However, defining a character limit as a threshold for adding a long description cannot be done across all languages and technologies:

1. Some languages need fewer characters to express the same information and defining a character limit (plus or minus x) for each language in use on the web is not really feasible.
2. Not all formats or technologies distinguish between a "short text alternative" (like HTML's `alt` attribute) and a long description. For example, Microsoft Word and PowerPoint used to have the fields "title" and "description" for text alternatives of images, but since Office 2019 (if I'm not mistaken) these have been replaced with a single field [`*`]. That doesn't mean long descriptions can't be provided in addition to the short text alternative, but the connection between the non-text object and the long description won't always be programmatically determinable. (E.g. in PowerPoint: putting the long description in a text container that is visually hidden behind an image that fills most of the slide.)

So the question becomes: how should we word guidance to authors to help them decide when to complement a short text alternative with a long description? Any character limits given in such guidance should be explicitly marked as language-specific and approximate.

---

`*` The title field did not provide a text alternative, but if you fill in both title and description for an image in PowerPoint 2016 and then open the presentation in PowerPoint 2019 or a newer version, the title and the description are combined into the Alt Text field. (Probably because many people mistakenly used Title as a text alternative in older versions of PowerPoint.)

**Comment by GreggVan** [comment]
+1
gregg

———————————
Professor, University of Maryland, College Park
Founder and Director Emeritus , Trace R&D Center, UMD
Co-Founder Raising the Floor. http://raisingthefloor.org
The Global Public Inclusive Infrastructure (GPII) http://GPII.net
The Morphic project  https://morphic.org

**Comment by cdesrosiers813** [comment]
Thanks for providing the history and next steps for discussion on this topic. The question on the table is: 

**How should we word guidance to authors to help them decide when to complement a short text alternative with a long description?** 

More guidance on character limits could help create more consistency across providers, as well as help authors understand image description best practices. I agree that the guidance should be noted as language specific. 

To kick off the discussion on this topic, how about providing a character limit that is slightly higher than what is popular in the field with a target average? For example, 250 character limit with a target of 125 characters. In this case, any image that requires a description over 250 characters should have a short + long description. 

What are your thoughts on this approach? Are there any suggested alternative options to explore?

**Comment by bruce-usab** [comment]
@cdesrosiers813 &ndash; this is in the context of wcag3, correct?  For me, a hard character limit makes less sense than ever.  Whatever the number chosen will be arbitrary, and 3 aims to be more research-based than 2.

My suggestion is for 3 to have _two_ **distinct** requirements for non-text context.
1. Short text alternative that provides descriptive identification.
2. Long text alternative that serves the equivalent purpose.

For most web technologies, the alt attribute is probably not the best choice for (2).

Having the requirements combined as they are in 1.1.1 has significant advantages (e.g., when null alt is best, or when the text equivalent is a name).  OTOH `alt` continues to be a topic of much discussion, so certainly there is the opportunity with 3 to communicate principles better.

**Comment by GreggVan** [comment]
Agree with Bruce

And where there IS no long description mechanism — then there certainly should not be a char limit.   Else how do you convey the information in a chart or diagram that is not otherwise in the text - just in the illustration. 

gregg

———————————
Professor, University of Maryland, College Park
Founder and Director Emeritus , Trace R&D Center, UMD
Co-Founder Raising the Floor. http://raisingthefloor.org
The Global Public Inclusive Infrastructure (GPII) http://GPII.net
The Morphic project  https://morphic.org

**Comment by cdesrosiers813** [comment]
Thanks for the great feedback. A few follow up questions:

@bruce-usab - How short should a "short text alternative" be? Are you saying it should be long enough to provide descriptive identification? I can see there being some confusion on "how short" and "descriptive identification" here. Perhaps this is the sort of guidance on principles you are suggesting, but I'm curious to hear your thoughts on these points. 

@GreggVan - What about when the provider does support long description? Do you think they should still avoid setting a character limit for alt text?

**Comment by patrickhlauke** [comment]
> @bruce-usab - How short should a "short text alternative" be? Are you saying it should be long enough to provide descriptive identification? I can see there being some confusion on "how short" and "descriptive identification" here. Perhaps this is the sort of guidance on principles you are suggesting, but I'm curious to hear your thoughts on these points.

The problem is of course that this is highly context specific and subjective. “Everything should be as simple as it can be, but not simpler”
Binary WCAG 2.x can't set a hard limit or give an unambiguous definition of how short "short" is, in a universal and testable way. The best it can hope for is to provide some non-binding, non-normative suggestions.

**Comment by GreggVan** [comment]
> Thanks for the great feedback. A few follow up questions:
> 
> @GreggVan <https://github.com/GreggVan> - What about when the provider does support long description? Do you think they should still avoid setting a character limit for alt text?
> 

    GV: if long description is supported by BOTH the technology being used and by assistive technologies — then ALT should be short and the long description used.

But for technologies  (including HTML and Powerpoint and ….) where there is no long description ALT text may need to be longer. 

  (Thankfully ARIA-Description <https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-description> was added to provide long description capability for HTML.  A thank you to all that made that happen). 

   


gregg

———————————
Professor, University of Maryland, College Park
Founder and Director Emeritus , Trace R&D Center, UMD
Co-Founder Raising the Floor. http://raisingthefloor.org
The Global Public Inclusive Infrastructure (GPII) http://GPII.net
The Morphic project  https://morphic.org

**Comment by ghost** [comment]
IHMO a character limit is a valid option to provide more acc to users of braille-interfaces. Reading an long alttext can be really annoying if its out bounce for the length of the braille interface. Therefore a max length should be based on the standard sizes of thes interfaces. Sizes differ on the model of the interface. So devs should be able to choose the best for there specific use case. Providing Options from 12 40 and 80 should be appropriate.

**Comment by patrickhlauke** [comment]
But again, as a best practice advice, sure. But not as a hard pass/fail criterion, as the point of WCAG is not to abolish things that are "annoying"...

**Comment by bruce-usab** [comment]
If anyone has data where the early limits came from, that might be instructive.  For example, IE at one point used ALT for mouse-over hover text but only so many characters.  Did that behavior become enshrined as the canonical limit?

I am inclined to tag this as wontfix since where should this issue go?

**Comment by cstrobbe** [comment]
> If anyone has data where the early limits came from, that might be instructive. 

I checked the old web accessibility books I have, and here are the results:

* *Constructing Accessible Web Sites* by Jim Thatcher et al (Glasshaus, 2002): no character limit is mentioned; just advice to keep alt text short (page 77).
* *Maximum Accessibility* by John Slatin and Sharron Rush (Addison-Wesley, 2003): no character limit is mentioned (pages 249–250).
* *Building Accessible Websites* by Joe Clark (New Riders, 2003): page 63 (emphasis mine):  
   > Keep the alt text short. There is no set limit on the length of an alt text, but as we shall see shortly, a very long alt may not be fully displayed when image-loading is turned off or when the browser cannot locate the image file. **By convention, limit alt texts to 1,024 characters (1 K) or less.**
* *Accessible XHTML and CSS Web Sites* by Jon Duckett (Wrox / Wiley, 2005): nothing about keeping alt text concise, let alone a character limit (pages 259–260).
* *Web Accessibility: Web Standards and Regulatory Compliance* by Jim Thatcher et al (Friends of ED, 2006): no character limit is mentioned; just advice to keep alt text short (pages 136, 141).
* *Access by Design* by Sarah Horton (New Riders, 2006): no character limit is mentioned (pages 78–80).
* *Design Accessible Websites* by Jeremy J. Sydik (The Pragmatic Programmers, 2007): page 159: "Most of the time, the alt text should be no more than 40–80 characters or a couple of seconds when spoken." This does not sound like a hard character limit.
* *Web Accessibility: Practical Advice for the Library and Information Professional*, edited by Jenny Craven (Facet Publishing, 2008): nothing on alt text length (page 33).
* *Universal Design for Web Applications* by Wendy Chisholm and Matt May (O'Reilly, 2009): no character limit is mentioned (pages 26–29).
* *Barrierefreiheit verstehen und umsetzten* by Jan Eric Hellbusch and Kerstin Probiesch (dpunkt.verlag, 2011): no character limit is mentioned (pages 157–158, 379–380, 389).
* *Accessibilité Web. Normes et bonnes pratiques pour des sites plus accessibles* by Armony Altinier (Eyrolles, 2012): no character limit is mentioned on pages 90–92, but page 96 mentions that the now defunct AccessiWeb guide "recommended" at most 80 characters as a guideline, i.e. without defining this as a hard limit. However, page 97 says that if more than 80 characters are needed, the image requires a long description.

Only two of the above books mentions a limit: Joe Clark defines 1,024 characters as the upper limit (without recommending that you go that far), which is much higher than the 150 or 125 characters that many blog posts claim. Armony Altinier gives 80 characters (for French content, at least) as a kind of rough cut-off value between a short alt text and images that also require a long description. But this is not worded as a hard limit.

It seems there have been versions of JAWS that cut off alt text after a certain number of characters. See for example the forum thread [Alt text character limit?](https://community.articulate.com/discussions/discuss/alt-text-character-limit/324467) on E-Learning Heroes from 2018, which contains the following statements:

> I'm having the same issue. Alt-text entered into Storyline for an image is 71 words (430 characters including spaces). JAWS reads 38 words (232 characters including spaces) and stops.

> JAWS stops reading after a certain number of characters and then requires the user to press the down arrow to continue reading. You can customize the number of characters read by changing the Maximum Line Length in the Settings Center.

The JAWS versions are not identified. 

Terrill Thompson's [ALT Length Test](https://terrillthompson.com/tests/altlength.html) (undated, but referenced by Eric Eggert's short blog post [here is no character limit for “alt text”: Myth Debunked!](https://yatil.net/blog/there-is-no-character-limit-for-alt-text), 19 February 2022) mentions the following type of "limit":

> JAWS 6.0 - reads ALT text in 125 character chunks (excluding spaces). If ALT text exceeds 125 characters, it divides it into multiple graphics, each having no more than 125 characters of ALT text, beginning where the previous graphic left off.

JAWS 6 was released [20 years ago](https://www.freelists.org/post/jaws-uk/IMPORTANT-JAWS-6-RELEASE-INFORMATION). I assume screen readers have moved on since then ...

**Comment by MarcHaunschild** [comment]
I don't understand this discussion. There is no SC that demands to put anything into alt attribute.
We have to provide an alternative and it's completely fine if it's visible.

I would even recommend to describe pictures with text, that is accessible for all. Why hide explanations from any person without assistive technology? That makes no sense (yet I regularly find examples of definitely important informations exclusively provided in alt attribute during my daily testing sessions).

Just keep the text in the alt attribute brief. If you need more than a few words to describe the image, this text probably is important not only for users with screen readers.

You still can work with disclosures (detail and summary) if you don't want to show it permanently.

Although I think a generous limit of 256 characters could help to make more people go for better solutions, the don't see the need. WCAG allows poor design and difficult to use websites. It doesn't say make it easy to use. It just says make it operable somehow...

**Reply by cdesrosiers813** [reply]
Thanks for pinging this thread, Marc. I have been meaning to return to it. 

I originally started this conversation because of questions I was receiving from Scribely's clients. They were wondering whether the W3C recommended a character limit for alt text as a best practice, and when they should consider providing an extended description AKA long description. Four years later, I'm still receiving this question. 

In addition to this confusion about what "short alt text means," technology providers are not creating solutions to help users implement extended descriptions. Providers seem willing to increase the character limit of alt text but they are not building solutions to support the method of keeping alt text short with a link to more information. The result is long unstructured descriptions in alt attributes which can be annoying for users, as well as confusing extended description workarounds which are difficult for publishers. 

I still think it would be good to discuss a best practice to clear up some of this confusion for publishers and providers about "how short is short" for alt text and when an extended description is needed. It would help to have a definitive answer or at least best practice recommendation from the primary source to encourage the development of solutions to this ongoing issue. 

Does anyone have suggestions on how we might reach a consensus to resolve this thread? Happy to participate in a discussion to resolve it. Please let me know how I can best support reaching a resolution.
